File: example-implementations/go/RingBuffer/RingBuffer.go
package ringbuffer

type RingBuffer[T any] struct {
	length   int
	capacity int
	head     int
	tail     int
	data     []T
}

func (r *RingBuffer[T]) Len() int { return r.length }

func NewRingBuffer[T any](capacity int) *RingBuffer[T] {
	return &RingBuffer[T]{
		length:   0,
		capacity: capacity,
		head:     0,
		tail:     0,
		data:     make([]T, capacity),
	}
}

func (r *RingBuffer[T]) Push(item T) {
	r.data[r.tail] = item
	r.tail = (r.tail + 1) % r.capacity

	if r.length < r.capacity {
		r.length++
	} else {
		r.head = (r.head + 1) % r.capacity
	}
}

func (r *RingBuffer[T]) Pop() (T, bool) {
	var zero T
	if r.length == 0 {
		return zero, false
	}

	val := r.data[r.head]
	r.head = (r.head + 1) % r.capacity
	r.length--

	return val, true
}

func (r *RingBuffer[T]) Get(idx int) (T, bool) {
	var zero T
	if idx < 0 || idx >= r.length {
		return zero, false
	}

	actualIdx := (r.head + idx) % r.capacity
	return r.data[actualIdx], true
}

func (r *RingBuffer[T]) IsFull() bool {
	return r.length == r.capacity
}

func (r *RingBuffer[T]) IsEmpty() bool {
	return r.length == 0
}

File: example-implementations/python/RingBuffer/ring_buffer.py
"""
Ring Buffer (Circular Buffer) implementation.
Fixed-size buffer that wraps around when full.
"""

from typing import TypeVar, Generic, Optional, List, Any

T = TypeVar('T')


class RingBuffer(Generic[T]):
    """Circular buffer with fixed capacity."""

    def __init__(self, capacity: int):
        self._length = 0
        self._capacity = capacity
        self._head = 0
        self._tail = 0
        self._data: List[Any] = [None] * capacity

    def __len__(self) -> int:
        return self._length

    def push(self, item: T) -> None:
        """Add item to buffer. Overwrites oldest item if full."""
        self._data[self._tail] = item
        self._tail = (self._tail + 1) % self._capacity

        if self._length < self._capacity:
            self._length += 1
        else:
            # Buffer is full, move head forward (overwrite oldest)
            self._head = (self._head + 1) % self._capacity

    def pop(self) -> Optional[T]:
        """Remove and return oldest item. Returns None if empty."""
        if self._length == 0:
            return None

        val = self._data[self._head]
        self._head = (self._head + 1) % self._capacity
        self._length -= 1

        return val

    def get(self, idx: int) -> Optional[T]:
        """Get item at logical index. Returns None if out of bounds."""
        if idx < 0 or idx >= self._length:
            return None

        actual_idx = (self._head + idx) % self._capacity
        return self._data[actual_idx]

    def is_full(self) -> bool:
        """Check if buffer is at capacity."""
        return self._length == self._capacity

    def is_empty(self) -> bool:
        """Check if buffer is empty."""
        return self._length == 0
